panic when aborting backend input

Interrupting an input request for a backend config results in a panic in the config shims

```
panic: interface conversion: interface {} is string, not map[string]interface {}

goroutine 1 [running]:
github.com/hashicorp/terraform/helper/schema.(*Backend).shimConfig(0xc0002bf260, 0x29fef20, 0xc0000c8110, 0x240cde0, 0x3986000, 0x29fef20)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/helper/schema/backend.go:185 +0xdb
github.com/hashicorp/terraform/helper/schema.(*Backend).PrepareConfig(0xc0002bf260, 0x29fef20, 0xc0000c80f0, 0x240cde0, 0x3986000, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/helper/schema/backend.go:133 +0x28c
github.com/hashicorp/terraform/command.(*Meta).backendInitFromConfig(0xc00049f6c0, 0xc0000d8140, 0xc0000a01be, 0xc00054ec48, 0x1035803, 0x0, 0xc00054ec58, 0x1034f3c, 0xc000000300, 0x300000002, ...)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/command/meta_backend.go:1043 +0x2bf
github.com/hashicorp/terraform/command.(*Meta).backend_C_r_s(0xc00049f6c0, 0xc0000d8140, 0x36fc0826, 0xc0000ca2a0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/command/meta_backend.go:613 +0x6f
github.com/hashicorp/terraform/command.(*Meta).backendFromConfig(0xc00049f6c0, 0xc0002be3c0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/command/meta_backend.go:462 +0xd68
github.com/hashicorp/terraform/command.(*Meta).Backend(0xc00049f6c0, 0xc0002be3c0, 0x0, 0xc00054f300, 0x1, 0x1, 0x0)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/command/meta_backend.go:85 +0x5d1
github.com/hashicorp/terraform/command.(*InitCommand).initBackend(0xc00049f6c0, 0xc0004749a0, 0x26cb1fe, 0xf, 0xc0002e9f60, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/command/init.go:450 +0x1ab
github.com/hashicorp/terraform/command.(*InitCommand).Run(0xc00049f6c0, 0xc0000cc010, 0x0, 0x0, 0xc0002ea130)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/command/init.go:287 +0x18ac
github.com/hashicorp/terraform/vendor/github.com/mitchellh/cli.(*CLI).Run(0xc00046e280, 0xc00046e280, 0xc0003ffd20, 0x1)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/vendor/github.com/mitchellh/cli/cli.go:255 +0x207
main.wrappedMain(0x0)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/main.go:223 +0xb37
main.realMain(0x0)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/main.go:100 +0xb7
main.main()
	/Users/jbardin/go/src/github.com/hashicorp/terraform/main.go:36 +0x2e
```

--- a/command/meta_backend.go
+++ b/command/meta_backend.go
@@ -1038,6 +1038,13 @@ func (m *Meta) backendInitFromConfig(c *configs.Backend) (backend.Backend, cty.V
 		if err != nil {
 			diags = diags.Append(fmt.Errorf("Error asking for input to configure backend %q: %s", c.Type, err))
 		}
+
+		// We get an unknown here if the if the user aborted input, but we can't
+		// turn that into a config value, so set it to null and let the provider
+		// handle it in PrepareConfig.
+		if !configVal.IsKnown() {
+			configVal = cty.NullVal(configVal.Type())
+		}
 	}
 
 	newVal, validateDiags := b.PrepareConfig(configVal)

--- a/helper/schema/backend.go
+++ b/helper/schema/backend.go
@@ -182,7 +182,11 @@ func (b *Backend) Configure(obj cty.Value) tfdiags.Diagnostics {
 // that should be populated enough to appease the not-yet-updated functionality
 // in this package. This should be removed once everything is updated.
 func (b *Backend) shimConfig(obj cty.Value) *terraform.ResourceConfig {
-	shimMap := hcl2shim.ConfigValueFromHCL2(obj).(map[string]interface{})
+	shimMap, ok := hcl2shim.ConfigValueFromHCL2(obj).(map[string]interface{})
+	if !ok {
+		// If the configVal was nil, we still want a non-nil map here.
+		shimMap = map[string]interface{}{}
+	}
 	return &terraform.ResourceConfig{
 		Config: shimMap,
 		Raw:    shimMap,

--- a/helper/schema/backend_test.go
+++ b/helper/schema/backend_test.go
@@ -32,6 +32,21 @@ func TestBackendPrepare(t *testing.T) {
 		},
 
 		{
+			"Null config",
+			&Backend{
+				Schema: map[string]*Schema{
+					"foo": &Schema{
+						Required: true,
+						Type:     TypeString,
+					},
+				},
+			},
+			nil,
+			map[string]cty.Value{},
+			true,
+		},
+
+		{
 			"Basic required field set",
 			&Backend{
 				Schema: map[string]*Schema{
@@ -111,7 +126,11 @@ func TestBackendPrepare(t *testing.T) {
 
 	for i, tc := range cases {
 		t.Run(fmt.Sprintf("%d-%s", i, tc.Name), func(t *testing.T) {
-			configVal, diags := tc.B.PrepareConfig(cty.ObjectVal(tc.Config))
+			cfgVal := cty.NullVal(cty.Object(map[string]cty.Type{}))
+			if tc.Config != nil {
+				cfgVal = cty.ObjectVal(tc.Config)
+			}
+			configVal, diags := tc.B.PrepareConfig(cfgVal)
 			if diags.HasErrors() != tc.Err {
 				for _, d := range diags {
 					t.Error(d.Description())

